commit b5f39d8be23d92ff292b1a109231e803222861d0 (refs/changes/94/1194/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-29T12:04:46-08:00 (2 years, 9 months ago)
    
    CNS-190 AXFR failure due to very large number of records on a single name

diff --git a/lib/dns-server.js b/lib/dns-server.js
index 5912027..6afecad 100644
--- a/lib/dns-server.js
+++ b/lib/dns-server.js
@@ -1131,15 +1131,20 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 					var name = names[i];
 					var recs = JSON.parse(zrecs[name]);
 					assert.arrayOfObject(recs);
-					self.addAnswers(q, recs,
-					    name + '.' + z);
-					pushed += recs.length;
-					if (pushed > MAX_ANS_PER_MSG) {
-						if (q.send() === false) {
-							finish();
-							return;
+					var trecs;
+					while (recs.length > 0) {
+						trecs = recs.slice(0, 20);
+						recs = recs.slice(20);
+						self.addAnswers(q, trecs,
+						    name + '.' + z);
+						pushed += trecs.length;
+						if (pushed > MAX_ANS_PER_MSG) {
+							if (!q.send()) {
+								finish();
+								return;
+							}
+							pushed = 0;
 						}
-						pushed = 0;
 					}
 				}
 				finish();
